File: scripts/render-review.py
"""Render seven geometrically different views of the same captured input. Pillow + ffmpeg.
Preview: native scene 0 RGB plus alternate production-renderer replays.
Video audio: original fixture only, never a commercial recording.
"""
import pathlib,json,subprocess,statistics,sys
from PIL import Image,ImageDraw,ImageFont
root=pathlib.Path(__file__).resolve().parent.parent
font='/System/Library/Fonts/STHeiti Light.ttc';fonts={}
def text(d,xy,s,size=18,fill='#acbcd0'):
 if size not in fonts:fonts[size]=ImageFont.truetype(font,size)
 d.text(xy,s,font=fonts[size],fill=fill)
def grid(d,a,x,y,w):
 step=w/8
 for py in range(8):
  for px in range(8):
   k=((7-py)*8+px)*3;c=tuple(round(255*(b+(v/127)**.7*f)) for b,v,f in zip([.014,.021,.032],a[k:k+3],[.9,.92,.94]))
   d.rounded_rectangle((x+px*step,y+py*step,x+(px+1)*step-4,y+(py+1)*step-4),3,fill=c,outline='#172131')
titles=['重力／沉積','天體／公轉','織光／經緯','門廊／縱深','雙生／呼應','拼光／碎片','聲場／三域']
palettes=['琥珀冰川','月夜紫羅蘭','翡翠珊瑚','鈷藍熔岩','蘭花青檸','桃紅電光','琥珀冰川']
selected=json.loads((root/'tmp/visual-review-selected.json').read_text());im=Image.new('RGB',(1200,1430),'#080e18');d=ImageDraw.Draw(im)
text(d,(32,22),'LUNAR LENS  1.4.0',30,'#edf4fc');text(d,(32,66),'同一音訊，同一時刻。七種空間構成，各自保留五個聲音行為。',18)
for i,t in enumerate(titles):
 x=32+(i%3)*396;y=112+(i//3)*330
 text(d,(x,y),f'0{i+1}  '+t,21,'#d6e4f2');text(d,(x+216,y+4),palettes[i],14);grid(d,selected['grids'][i],x,y+32,280)
text(d,(32,1115),'同一幀，逐層獨看',23,'#edf4fc')
for i,t in enumerate(['衝擊 IMPACT','地基 BODY','前景 PHRASE','鋪陳 BED','細節 DETAIL']):
 x=32+i*235;text(d,(x,1165),t,17);grid(d,selected['grids'][7+i],x,1198,184)
text(d,(32,1400),selected['caption']+f" · {selected['position']:.2f} s · Native 特徵／正式渲染器重播 · 螢幕 gamma 預覽",14)
im.save(root/'media/preview.png')
if '--sheet-only' in sys.argv:print('Wrote preview.png.');sys.exit(0)
rows=json.loads((root/'tmp/visual-review-frames.json').read_text());origin=statistics.median(r['time']-r['position'] for r in rows if 1<r['position']<44)-.05
fps=30;start=26;duration=17;idx=0;silent=root/'tmp/review-silent.mp4'
proc=subprocess.Popen(['ffmpeg','-y','-v','error','-f','rawvideo','-pix_fmt','rgb24','-s','1200x1170','-r',str(fps),'-i','-','-an','-c:v','libx264','-crf','18','-pix_fmt','yuv420p',str(silent)],stdin=subprocess.PIPE)
for n in range(fps*duration):
 t=start+n/fps
 while idx+1<len(rows) and rows[idx+1]['time']<=origin+t:idx+=1
 row=rows[idx];im=Image.new('RGB',(1200,1170),'#080e18');d=ImageDraw.Draw(im)
 text(d,(32,20),'LUNAR LENS  1.4.0 / 同一聲音，七種空間',27,'#edf4fc')
 label='低頻＋短和弦＋打擊' if t<34 else '移除低頻：剩中頻短和弦' if t<38 else '只剩高頻短音' if t<42 else '停止後退暗'
 text(d,(32,62),label,18);text(d,(1040,26),f'{t:05.2f} s',21)
 for i,name in enumerate(titles):
  x=32+(i%3)*396;y=114+(i//3)*336;text(d,(x,y-4),name,19);grid(d,row['grids'][i],x,y+28,276)
 text(d,(32,1130),'原創驗收訊號 · 相同 Native 分析幀 · 正式渲染器重播 · 非實機 LED 拍攝',15)
 proc.stdin.write(im.tobytes())
proc.stdin.close();assert proc.wait()==0
subprocess.run(['ffmpeg','-y','-v','error','-i',str(silent),'-ss',str(start),'-i',str(root/'tmp/role-fixture.wav'),'-t',str(duration),'-map','0:v','-map','1:a','-af','volume=0.4','-c:v','copy','-c:a','aac','-b:a','192k','-movflags','+faststart',str(root/'media/roles-in-motion.mp4')],check=True)
print('Wrote preview.png and 17-second seven-composition movie.')

File: scripts/render-review.mjs
// Same captured native features, replayed through all six production compositions.
import fs from 'node:fs';import path from 'node:path';import {fileURLToPath} from 'node:url';import {createRequire} from 'node:module';
const root=path.resolve(path.dirname(fileURLToPath(import.meta.url)),'..'),require=createRequire(import.meta.url),V=require('../code/lens_visual.js');
const source=process.argv[2]||'tmp/role-fixture-native.jsonl',caption=process.argv[3]||'原創混合驗收訊號';
const rows=fs.readFileSync(path.join(root,source),'utf8').trim().split('\n').map(JSON.parse),engines=Array.from({length:11},()=>new V.Engine()),out=[];
let previous=rows[0].time,chosen,best=-1;
for(const row of rows){const dt=Math.max(.001,Math.min(.1,row.time-previous));previous=row.time;
 const grids=engines.map((e,i)=>e.step(row.features,dt,{...row.params,scene:i<6?i:0,palette:0,focus:i<6?0:i-5,freeze:false,black:false}));
 // Exact native RGB for the recorded default scene, replay for alternative views.
 if(row.params.scene===0&&row.params.focus===0&&!row.params.black&&!row.params.freeze)grids[0]=row.leds;
 const frame={time:row.time,position:row.position,roles:row.features.roles,grids,running:row.running,caption};out.push(frame);
 const r=row.features.roles,score=r[0]*r[1]*r[2]*(.15+r[3])*(.15+r[4]);if(row.running&&row.position>1&&score>best){best=score;chosen=frame;}
}
fs.writeFileSync(path.join(root,'tmp/visual-review-frames.json'),JSON.stringify(out));fs.writeFileSync(path.join(root,'tmp/visual-review-selected.json'),JSON.stringify(chosen));
const text=(x,y,s,size=15,col='#a8b8cb')=>`<text x="${x}" y="${y}" fill="${col}" font-size="${size}">${s}</text>`;
function grid(a,x,y,w){let s='',step=w/8;for(let py=0;py<8;py++)for(let px=0;px<8;px++){const k=((7-py)*8+px)*3,c=a.slice(k,k+3).map((v,i)=>Math.round(255*([.014,.021,.032][i]+Math.pow(v/127,.7)*[.9,.92,.94][i])));s+=`<rect x="${x+px*step}" y="${y+py*step}" width="${step-4}" height="${step-4}" rx="3" fill="rgb(${c.join(',')})" stroke="#172131" stroke-width=".6"/>`;}return s;}
let svg='<svg xmlns="http://www.w3.org/2000/svg" width="1200" height="1100"><rect width="1200" height="1100" fill="#080e18"/><g font-family="Arial,Hiragino Sans GB,sans-serif">';
svg+=text(32,48,'LUNAR LENS  1.2.1',28,'#edf4fc')+text(32,78,'同一音訊，同一時刻。六種空間構成，各自保留五個聲音行為。',18);
V.names.forEach((t,i)=>{const x=32+(i%3)*396,y=122+Math.floor(i/3)*330;svg+=text(x,y,`0${i+1}  ${t}`,19,'#d6e4f2')+text(x+206,y,V.paletteNames[i],13)+grid(chosen.grids[i],x,y+22,280);});
svg+=text(32,810,'同一幀，逐層獨看',21,'#edf4fc');
['衝擊 IMPACT','地基 BODY','前景 PHRASE','鋪陳 BED','細節 DETAIL'].forEach((t,i)=>{const x=32+i*235;svg+=text(x,847,t,16)+grid(chosen.grids[6+i],x,868,184);});
svg+=text(32,1080,`${caption} · ${chosen.position.toFixed(2)} s · Native 特徵／正式渲染器重播 · 螢幕 gamma 預覽`,13)+'</g></svg>';
fs.writeFileSync(path.join(root,'media/preview.svg'),svg);console.log(`Prepared ${rows.length} frames; selected ${chosen.position.toFixed(2)} s.`);
